src/requirements/Farcaster/components/FarcasterCastHash.tsx

```diff
@@ -14,28 +17,65 @@ const FarcasterCastHash = ({ baseFieldPath }: Props) => {
     formState: { errors },
   } = useFormContext()
 
+  const hash = useWatch({ name: `${baseFieldPath}.data.hash` })
+  const url = useWatch({ name: `${baseFieldPath}.data.url` })
+  const debouncedHash = useDebouncedState(hash)
+  const debouncedUrl = useDebouncedState(url)
+
+  const { data, isLoading, error } = useFarcasterCast(debouncedHash, debouncedUrl)
+
   return (
-    <FormControl
-      isRequired
-      isInvalid={parseFromObject(errors, baseFieldPath)?.data?.hash}
-    >
-      <FormLabel>Cast hash:</FormLabel>
-
-      <Input
-        {...register(`${baseFieldPath}.data.hash`, {
-          required: "This field is required.",
-          pattern: {
-            value: ADDRESS_REGEX,
-            message:
-              "Please input a 42 characters long, 0x-prefixed hexadecimal hash.",
-          },
-        })}
-      />
-
-      <FormErrorMessage>
-        {parseFromObject(errors, baseFieldPath)?.data?.hash?.message}
-      </FormErrorMessage>
-    </FormControl>
+    <>
+      {(!!data || !!isLoading || !!error) && (
+        <FarcasterCast cast={data} loading={isLoading} error={!!error} />
+      )}
+
+      <FormControl
+        isRequired={!hash}
+        isInvalid={!hash && parseFromObject(errors, baseFieldPath)?.data?.url}
+      >
+        <FormLabel opacity={!!hash ? 0.3 : 1}>Cast URL:</FormLabel>
+
+        <Input
+          {...register(`${baseFieldPath}.data.url`, {
+            required: !hash ? "This field is required." : false,
+            disabled: !!hash,
+            pattern: {
+              value: /^https:\/\/(.)+\.(.)+$/,
+              message: "Invalid URL",
+            },
+          })}
+        />
+
+        <FormErrorMessage>
+          {!hash && parseFromObject(errors, baseFieldPath)?.data?.url?.message}
+        </FormErrorMessage>
+      </FormControl>
+
+      <FormControl
```

Comment:
Do we still need the hash input? If so, ideally we should merge the logics into one input so it works with both as `Cast URL or hash`. Now I find it confusing that at first it looks like I have to provide both. And we should display the FarcasterCast below that when filled, as it depends on it's value (it looks amazing btw!)

Reply:
Oh I see that we also save them differently and it'd be complicated to reverse fetch what to display. Can't we just always save the hash (I think that's safer than an url?) and use the url to get the hash and vica versa?

Reply:
I ended up implementing another solution, because supporting both hashes and urls seamlessly with 1 input required a pretty hacky solution. So, I added an uncontrolled input for the URL: if the user pastes a valid URL & we can fetch the hash, we'll set `data.hash` to the fetched one, otherwise we just display an error.